packager/binarypackaging.py

```diff
@@ -0,0 +1,106 @@
+                        with zipfile.ZipFile(zips[0], 'r') as zf:
+                            try:
+                                tree = ET.fromstring(zf.read(os.path.join(addon_id.split('+')[0], 'addon.xml')))
+                                version = tree.attrib['version'].encode('utf-8')
+
+                                imports = [(elem.attrib['addon'], LooseVersion(elem.attrib.get('version', '0.0.0')))
+                                           for elem in tree.findall('./requires/import')]
+                                if meets_version_requirements(imports, min_versions):
+                                    yield Artifact(addon_id.split('+')[0], version, zips[0], addon_id.split('+')[1])
+                            except (ET.ParseError, KeyError, IndexError) as e:
+                                logging.exception("Failed to read addon info from '%s'. Skipping" % archive)
+                                continue
+
+
+def filter_latest_version(artifacts):
+    artifacts = list(artifacts)
+    artifacts.sort(key=lambda _: _.addon_id)
+    for addon_id, versions in groupby(artifacts, key=lambda _: _.addon_id + _.treeish):
+        versions = list(versions)
+        versions.sort(key=lambda _: LooseVersion(_.version), reverse=True)
+        if len(versions) >= 2 and versions[0].version == versions[1].version:
+            logger.warning("Duplicate artifact. I will pick one at random. First: %r. Second: %r", versions[0], versions[1])
+        yield versions[0]
+
+
+def write_artifact(artifact, outdir):
+    # Reads artifact data from git and writes a zip file (and companion files) to `outdir`
+    with utils.tempdir() as unpack_dir:
+        with zipfile.ZipFile(artifact.git_repo, 'r') as zf:
+            zf.extractall(unpack_dir)
+
+        with utils.tempdir() as package_dir:
+            pack_artifact(artifact, os.path.join(unpack_dir, artifact.addon_id), package_dir)
+            delete_companion_files(outdir)
+            copy_tree(package_dir, outdir)
+
+
+def update_changed_artifacts(binary_repos, refs, min_versions, outdir):
+    """ Returns a tuple with number of new and a list of all artifacts. """
+
+    artifacts = list(collect_artifacts(binary_repos, refs, min_versions))
+    artifacts = list(filter_latest_version(artifacts))
+
+    added = [a for a in artifacts if not os.path.exists(
+        os.path.join(outdir, "%s+%s" % (a.addon_id, a.treeish), "%s-%s.zip" % (a.addon_id, a.version)))]
+
+    for artifact in added:
+        logger.debug("New artifact %s+%s version %s", artifact.addon_id, artifact.treeish, artifact.version)
+        dest = os.path.join(outdir, "%s+%s" % (artifact.addon_id, artifact.treeish))
+        print(dest)
+        makedirs_ignore_errors(dest)
+        try:
+            write_artifact(artifact, dest)
+        except Exception as ex:
+            logger.error("Failed to package %s:", artifact, exc_info=1)
+
+    return len(added), ['%s+%s' % (a.addon_id, a.treeish) for a in artifacts]
```

Comment:
Why is treeish used here? That's git stuff used to know where in the git tree the addon exists

Reply:
I just wanted to reuse the already existing `Artifact` definition. Actually `treeish` is platform in this case.
I'll fix it that `Artifact` isn't a general type, but specific for the packaging method.